Integrate Gravatar Quick Editor

pinarol opened the pull request:
Fixes #

Integrates the Gravatar Quick Editor into Jetpack.
Changes are behind the remote feature flag: `gravatar_quick_editor`

<img src="https://github.com/user-attachments/assets/916ae964-2591-4022-9cf6-f9cd9a37691a" width=300/>

https://github.com/user-attachments/assets/a5baef59-fe18-4ff5-a5ec-211a156bbf21

https://github.com/user-attachments/assets/09521681-f8cc-45f5-8c52-ec4cc901dd5a

**Technical Details**

The new structure depends on the Quick Editor's update callback and fires a notification `GravatarQEAvatarUpdateNotification`. The UI components that render the avatar subscribe to that notification. Once received, they fetch the latest version of the avatar from the BE. To achieve this we introduce a way to `forceRefresh` the avatar, which wasn't present before. To achieve `forceRefresh` we add a cache buster query parameter otherwise the backend cache gets in the way and the latest version of the avatar can't be fetched.

**To test:**

**Enable the feature flag** "Gravatar Quick Editor" in `Me > App Settings > Debug > Feature Flags` (It's enabled by default in DEBUG config)
Log out from the Jetpack app
Log in with a new email (Highly recommend getting a temp email via https://tempmail.so/us , you should go to your inbox in your mobile device since the login depends on a magic link)
Use the email magic link to continue to the app
Observe: You landed on the signup epilogue, the avatar placeholder has a + on it and tapping on the avatar opens the Quick Editor.
Upload & change the avatar via the Quick Editor
Observe: The avatar on the signup page also gets updated
Close the Quick Editor
Continue to the app

Go to `Me > My Profile`
Tap either on the avatar or on the "Update profile photo" button
Observe: Quick Editor opens
Upload & change the avatar via the Quick Editor
Observe: The avatars on the main app gets updated as well (the one on the profile header, and the "Me" tab bar item at the bottom)

**Appearance Test**
Go to Me > App Settings > Appearance
Switch the appearance
Go to `Me > My Profile`
Tap either on the avatar or on the "Update profile photo" button
Observe: Quick Editor opens and the appearance matched with the app.

**Disable the feature flag** "Gravatar Quick Editor" in `Me > App Settings > Debug > Feature Flags` 

Repeat the tests ☝️
This time instead of the Gravatar Quick Editor you should observe that the system photo picker appears and the selected image is uploaded & set as the avatar.

## Regression Notes
1. Potential unintended areas of impact
The existing upload flow

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested with the feature flag off

3. What automated tests I added (or what prevented me from doing so)
x

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [x] WordPress.com sites and self-hosted Jetpack sites.
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [x] High contrast.
- [x] VoiceOver.
- [x] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [x] Multi-tasking: Split view and Slide over. (iPad)

pinarol committed:
Add remote FF gravatar_quick_editor

pinarol committed:
Add forceRefresh option for image download methods

pinarol committed:
Force refresh on gravatar update notification

pinarol committed:
Add `GravatarQuickEditorPresenter`

pinarol committed:
Display QuickEditor instead of the avatar upload menu

pinarol committed:
Move listenForGravatarChanges to upper level
Add forceRefresh option

pinarol committed:
Add one more FF check

pinarol committed:
Adjust the obj-c call

pinarol committed:
Listen to changes on the MeHeaderView

pinarol committed:
Await the image download to update the cached image

pinarol committed:
Listen to gravatar changes on the signup epilogue

pinarol committed:
Add one more FF check

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pinarol committed:
Separate the notification name for the QE
Add email check when handling the notification

`GravatarQEAvatarUpdateNotification`

pinarol committed:
Merge branch 'trunk' into wppinar/integrate-gravatar-quick-editor

# Conflicts:
#	WordPress/Classes/Utility/BuildInformation/RemoteFeatureFlag.swift

pinarol committed:
Add a release note

pinarol committed:
Revert whitespace change

pinarol committed:
Merge branch 'trunk' into wppinar/integrate-gravatar-quick-editor

# Conflicts:
#	WordPress/WordPress.xcodeproj/project.pbxproj

pinarol committed:
Use overrideImageCache`

pinarol commented:
cc: @kean @crazytonyli

pinarol committed:
Fix the old avatar issue

pinarol committed:
Merge branch 'trunk' into wppinar/integrate-gravatar-quick-editor

pinarol committed:
Update unit test

crazytonyli commented on WordPress/Classes/Extensions/URL+Helpers.swift:
```diff
@@ -151,4 +151,12 @@ extension URL {
+    /// bypass the backend cache and get the latest image.
+    func appendingGravatarCacheBusterParam() -> URL {
+        var urlComponents = URLComponents(url: self, resolvingAgainstBaseURL: false)
+        urlComponents?.queryItems?.append(.init(name: "_", value: "\(NSDate().timeIntervalSince1970)"))
```
Do you mind adding a simple unit test for this function? This function does not work if there is no query in the original URL.

```
> URL(string: "https://wordpress.com")!.appendingGravatarCacheBusterParam()
$R0: Foundation.URL = "https://wordpress.com"
```

pinarol replied:
Sure thing! And I handled that case as well. Thanks for pointing out.

crazytonyli commented on WordPress/Classes/ViewRelated/Me/My Profile/Gravatar/GravatarQuickEditorPresenter.swift:
```diff
@@ -0,0 +1,46 @@
+                Task {
+                    // Purge the cache otherwise the old avatars remain around.
+                    await ImageDownloader.shared.clearURLSessionCache()
+                    await ImageDownloader.shared.clearMemoryCache()
```
Purging all cache entries seems a bit aggressive. Would only removing the cached image for `email` work?

pinarol replied:
Despite being connected to the same email, avatars scattered throughout the app have usually different URLs depending on the query parameters like size, default image option, rating etc.

And unfortunately URLCache or NSCache doesn't give us chance to iterate over the cache keys and filter based on host and path. We need to know the exact URL to remove an entry which we don't.

Alternatively, we can remove these calls altogether but this causes some avatars to remain old in different places of the app which looks inconsistent and faulty. Also, killing & reopening the app can bring in the old avatar since URLCache has a disk cache. So I recommend keeping these purge operations but let me know if you prefer otherwise.

crazytonyli commented on WordPressAuthenticator/Sources/Signin/LoginLinkRequestViewController.swift:
```diff
@@ -38,18 +38,31 @@ class LoginLinkRequestViewController: LoginViewController {
         } else {
             gravatarView?.isHidden = true
         }
+        NotificationCenter.default.addObserver(self, selector: #selector(refreshAvatar), name: .GravatarQEAvatarUpdateNotification, object: nil)
```
`viewWillAppear` may get called multiple times. I don't think this line is meant to be called multiple times?

pinarol replied:
Oh I totally missed it, moved to viewDidLoad now. Thanks!

crazytonyli commented on WordPressAuthenticator/Sources/Unified Auth/View Related/Reusable Views/GravatarEmailTableViewCell.swift:
```diff
@@ -53,6 +58,20 @@ class GravatarEmailTableViewCell: UITableViewCell {
         containerView.layer.cornerRadius = hasBorders ? 8 : 0
         containerView.layer.borderColor = hasBorders ? UIColor.systemGray3.cgColor : UIColor.clear.cgColor
+
+        NotificationCenter.default.addObserver(self, selector: #selector(refreshAvatar), name: .GravatarQEAvatarUpdateNotification, object: nil)
```
This observer will also be added multiple times, when the cell is reused.

pinarol replied:
Right, moved to `init`. 👍

pinarol committed:
Move `addObserver` to viewDidLoad

pinarol committed:
Mode addObserver to init`

pinarol committed:
Add unit tests for `appendingGravatarCacheBusterParam` and handle the canonical URL as well

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/303">25.6</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pinarol committed:
Merge branch 'trunk' into wppinar/integrate-gravatar-quick-editor

# Conflicts:
#	WordPress/Classes/Utility/BuildInformation/RemoteFeatureFlag.swift
#	WordPress/WordPress.xcodeproj/project.pbxproj

pinarol reviewed (commented):
Thanks for the review @crazytonyli 🙇  This is ready for another look. (I was AFK last week so this is a bit of a late response)

crazytonyli commented on WordPress/Classes/ViewRelated/Gravatar/UIImageView+Gravatar.swift:
```diff
@@ -56,14 +58,31 @@ extension UIImageView {
+                    }
+                }
+            }
+
```
I'm not so sure about these changes. @kean Can you have a look please?

kean replied:
The app clears caches on `onAvatarUpdated` and they will be automatically re-populated later. I'm not sure it's worth adding complexity and making `task` getter `internal` for this scenario.

pinarol replied:
Normally we shouldn't need this yes. But unfortunately there's a delay at the backend that causes the old avatar to return for like 25 seconds after selecting a different avatar. So the old avatar comes back on the next request that doesn't have the cache buster parameter and looks like this:

<img src="https://github.com/user-attachments/assets/cf7fe31c-1d6d-4584-bab0-8f190a55e3e4" width=250/>

I reported this issue to systems and following it. In the meantime I update the memory cache for the cache buster-less version of the URL so the correct image can be shown next time.

kean replied:
Yeah, this doesn't look great, especially since it replaced the new avatar with the old one. I don't see anything particular wrong with adding this bit of code, so I'm going to leave it up to you.

crazytonyli commented:
@pinarol Thanks for addressing my comments. They look good to me. I'm not familiar with the image downloading implementation, and asked Alex to have to a look. ⬆️

kean commented on WordPress/Classes/ViewRelated/Me/My Profile/MyProfileHeaderView.swift:
```diff
@@ -25,7 +25,7 @@ class MyProfileHeaderView: UITableViewHeaderFooterView {
         didSet {
             if let email = gravatarEmail {
-                gravatarImageView.downloadGravatar(for: email, gravatarRating: .x)
+                downloadAvatar()
```
It produces a new Xcode warning (email unused).

pinarol replied:
Oh I missed this. Fixed. 👍

kean reviewed (approved):
LGTM with a couple of minor comments.

Some additional feedback:
- Clearing all caches isn't ideal but seems acceptable for this scenario. Ideally, it would use a separate cache and/or would rely on HTTP caching.
- (UX) It's a bit odd that after you upload a new image, it doesn't change your profile picture.
- There is no way to delete an avatar (pretty sure it is also the case in production)

pinarol committed:
Fix "unused var" warning

pinarol committed:
Merge branch 'trunk' into wppinar/integrate-gravatar-quick-editor

pinarol commented:
Thanks for the review @kean !

> (UX) It's a bit odd that after you upload a new image, it doesn't change your profile picture.

Yeah this is the same with web's Avatar Quick Editor as well as in gravatar.com, and a consistent UX was targeted. If no avatar is selected though, the uploaded avatar will be set immediately. But if there's already a selected avatar, the user needs to deliberately change the selection. A kind of precaution to prevent unintended changes to the user's global avatar.

> There is no way to delete an avatar (pretty sure it is also the case in production)

This is coming in a couple of months. For now, avatars can be deleted via gravatar.com only. We are waiting for the corresponding REST endpoint and we'll be adding this feature to the QE along with some other improvements.

pinarol committed:
Update release notes

pinarol committed:
Revert "Update release notes"

This reverts commit f0b3eb971980ad3f918b2c28bfdf8cebfefb518b.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/303">25.6</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

kean commented:
> But if there's already a selected avatar, the user needs to deliberately change the selection. A kind of precaution to prevent unintended changes to the user's global avatar.

It just it appeared broken to me. You can change it back is a single click, so does it need a precaution? If it's not a precaution, it's not clear what user scenario it solves by making a two-step process.

> For now, avatars can be deleted via gravatar.com only. 

It's be nice if it also includes removing your current avatar as I couldn't find a way to do it.

pinarol commented:
> It just it appeared broken to me. You can change it back is a single click, so does it need a precaution? If it's not a precaution, it's not clear what user scenario it solves by making a two-step process.

Let me take this feedback to our designer. I know that we want a unified experience between web/iOS/Android so I can't change this for iOS only but I am sure we can consider and discuss this. 

> It's be nice if it also includes removing your current avatar as I couldn't find a way to do it.

You mean on the web? https://gravatar.com/profile/avatars 
Because I can do it by removing the selected avatar:

<img width="384" alt="Screenshot 2024-11-15 at 16 02 35" src="https://github.com/user-attachments/assets/8ad931a3-5ab6-45dd-8355-44f6a6039869">

kean commented:
> You mean on the web? https://gravatar.com/profile/avatars. Because I can do it by removing the selected avatar:

Yeah, it would be nice to add soon too, but I think it needs a separate way to "remove current avatar" though without actually deleting it.

Overall, the new experience looks pretty great.
